File: tools/openadp-encrypt.py
#!/usr/bin/env python3
"""
OpenADP File Encryption Tool - Python Version

This tool exactly matches the functionality of cmd/openadp-encrypt/main.go
"""

import argparse
import os
import sys
import json
import getpass
import struct
import socket
from pathlib import Path
from Crypto.Cipher import AES
from Crypto.Random import get_random_bytes

# Add the openadp package to the path
sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'sdk', 'python'))

from openadp.keygen import generate_encryption_key, Identity
from openadp.client import get_servers, get_fallback_server_info, ServerInfo, OpenADPClient

VERSION = "1.0.0"
NONCE_SIZE = 12  # AES-GCM nonce size


class Metadata:
    """Metadata stored with encrypted files"""
    def __init__(self, servers, threshold, version, auth_code, user_id):
        self.servers = servers
        self.threshold = threshold
        self.version = version
        self.auth_code = auth_code
        self.user_id = user_id
    
    def to_dict(self):
        return {
            "servers": self.servers,
            "threshold": self.threshold,
            "version": self.version,
            "auth_code": self.auth_code,
            "user_id": self.user_id
        }


def show_help():
    print("""OpenADP File Encryption Tool

USAGE:
    openadp-encrypt.py --file <filename> [OPTIONS]

OPTIONS:
    --file <path>          File to encrypt (required)
    --password <password>  Password for key derivation (will prompt if not provided)
    --user-id <id>         User ID for secret ownership (will prompt if not provided)
    --servers <urls>       Comma-separated list of server URLs (optional)
    --servers-url <url>    URL to scrape for server list (default: https://servers.openadp.org)
    --version              Show version information
    --help                 Show this help message

USER ID SECURITY:
    Your User ID uniquely identifies your secrets on the servers. It is critical that:
    • You use the same User ID for all your files
    • You keep your User ID private (anyone with it can overwrite your secrets)
    • You choose a unique User ID that others won't guess
    • You remember your User ID for future decryption

    You can set the OPENADP_USER_ID environment variable to avoid typing it repeatedly.

SERVER DISCOVERY:
    By default, the tool fetches the server list from servers.openadp.org/api/servers.json
    If the registry is unavailable, it falls back to hardcoded servers.
    Use -servers to specify your own server list and skip discovery.

EXAMPLES:
    # Encrypt a file using discovered servers (fetches from servers.openadp.org)
    openadp-encrypt.py --file document.txt

    # Encrypt using specific servers (skip discovery)
    openadp-encrypt.py --file document.txt --servers "https://server1.com,https://server2.com"

    # Use a different server registry
    openadp-encrypt.py --file document.txt --servers-url "https://my-registry.com"

    # Use environment variables to avoid prompts
    export OPENADP_PASSWORD="mypassword"
    export OPENADP_USER_ID="myuserid"
    openadp-encrypt.py --file document.txt

The encrypted file will be saved as <filename>.enc
""")


def get_hostname():
    """Get hostname for device identification"""
    try:
        return socket.gethostname()
    except:
        return "unknown"


def write_uint32_le(f, value):
    """Write a 32-bit unsigned integer in little-endian format"""
    f.write(struct.pack('<I', value))


def encrypt_file(input_filename, password, user_id, server_infos, servers_url):
    """Encrypt a file using OpenADP servers"""
    output_filename = input_filename + ".enc"
    
    # Generate encryption key using OpenADP with full distributed protocol
    print("🔄 Generating encryption key using OpenADP servers...")
    # Create Identity from filename, user_id, and hostname (matching old derive_identifiers behavior)
    identity = Identity(
        uid=user_id,
        did=get_hostname(),
        bid=f"file://{os.path.basename(input_filename)}"
    )
    result = generate_encryption_key(identity, password, 10, 0, server_infos)
    
    if result.error:
        raise Exception(f"failed to generate encryption key: {result.error}")
    
    # Extract information from the result
    enc_key = result.encryption_key
    auth_codes = result.auth_codes
    actual_server_urls = [server_info.url for server_info in result.server_infos]
    threshold = result.threshold
    
    print(f"🔑 Generated authentication codes for {len(auth_codes.server_auth_codes)} servers")
    print(f"🔑 Key generated successfully (UID={user_id}, DID={get_hostname()}, BID=file://{os.path.basename(input_filename)})")
    
    # Show which servers were actually used for key generation
    if len(actual_server_urls) > 0 and len(actual_server_urls) != len(server_infos):
        print(f"📋 Servers actually used for key generation ({len(actual_server_urls)}):")
        for i, url in enumerate(actual_server_urls, 1):
            print(f"   {i}. {url}")
    
    # Read input file
    try:
        with open(input_filename, 'rb') as f:
            plaintext = f.read()
    except Exception as e:
        raise Exception(f"failed to read input file: {e}")
    
    # Generate random nonce
    nonce = get_random_bytes(NONCE_SIZE)
    
    # Create metadata using the actual results from keygen
    metadata = Metadata(
        servers=actual_server_urls,
        threshold=threshold,
        version="1.0",
        auth_code=auth_codes.base_auth_code,
        user_id=user_id
    )
    
    metadata_json = json.dumps(metadata.to_dict()).encode('utf-8')
    
    # Encrypt the file using metadata as additional authenticated data
    cipher = AES.new(enc_key, AES.MODE_GCM, nonce=nonce)
    cipher.update(metadata_json)  # Additional authenticated data
    ciphertext, tag = cipher.encrypt_and_digest(plaintext)
    
    # Combine ciphertext and tag for compatibility with Go's GCM implementation
    encrypted_data = ciphertext + tag
    
    # Write encrypted file: [metadata_length][metadata][nonce][encrypted_data]
    try:
        with open(output_filename, 'wb') as f:
            # Write metadata length (4 bytes, little endian)
            write_uint32_le(f, len(metadata_json))
            
            # Write metadata
            f.write(metadata_json)
            
            # Write nonce
            f.write(nonce)
            
            # Write encrypted data
            f.write(encrypted_data)
    except Exception as e:
        raise Exception(f"failed to create output file: {e}")
    
    print(f"📁 Input:  {input_filename} ({len(plaintext)} bytes)")
    print(f"📁 Output: {output_filename} ({4 + len(metadata_json) + NONCE_SIZE + len(encrypted_data)} bytes)")
    print(f"🔐 Encryption: AES-GCM")
    print(f"🌐 Servers: {len(actual_server_urls)} servers used")
    print(f"🎯 Threshold: {threshold}-of-{len(actual_server_urls)} recovery")
    
    # Show final server list stored in metadata
    print(f"📋 Servers stored in encrypted file metadata:")
    for i, url in enumerate(actual_server_urls, 1):
        print(f"   {i}. {url}")


def main():
    parser = argparse.ArgumentParser(description="OpenADP File Encryption Tool", add_help=False)
    parser.add_argument("--file", dest="filename", help="File to encrypt (required)")
    parser.add_argument("--password", help="Password for key derivation (will prompt if not provided)")
    parser.add_argument("--user-id", dest="user_id", help="User ID for secret ownership (will prompt if not provided)")
    parser.add_argument("--servers", help="Comma-separated list of server URLs (optional)")
    parser.add_argument("--servers-url", dest="servers_url", default="https://servers.openadp.org", 
                       help="URL to scrape for server list (default: https://servers.openadp.org)")
    parser.add_argument("--version", action="store_true", help="Show version information")
    parser.add_argument("--help", action="store_true", help="Show help information")
    
    args = parser.parse_args()
    
    if args.version:
        print(f"OpenADP File Encryption Tool v{VERSION}")
        return
    
    if args.help:
        show_help()
        return
    
    if not args.filename:
        print("Error: --file is required")
        show_help()
        sys.exit(1)
    
    # Check if input file exists
    if not os.path.exists(args.filename):
        print(f"Error: Input file '{args.filename}' not found.")
        sys.exit(1)
    
    # Get password (priority: flag > environment > prompt)
    password_str = ""
    if args.password:
        password_str = args.password
        print("⚠️  Warning: Password provided via command line (visible in process list)")
    elif os.environ.get("OPENADP_PASSWORD"):
        password_str = os.environ.get("OPENADP_PASSWORD")
        print("Using password from environment variable")
    else:
        password_str = getpass.getpass("Enter password: ")
    
    # Get user ID (priority: flag > environment > prompt)
    user_id_str = ""
    if args.user_id:
        user_id_str = args.user_id
        print("⚠️  Warning: User ID provided via command line (visible in process list)")
    elif os.environ.get("OPENADP_USER_ID"):
        user_id_str = os.environ.get("OPENADP_USER_ID")
        print("Using user ID from environment variable")
    else:
        user_id_str = input("Enter your user ID (this identifies your secrets): ").strip()
        if not user_id_str:
            print("Error: User ID cannot be empty")
            sys.exit(1)
    
    # Validate user ID
    user_id_str = user_id_str.strip()
    if len(user_id_str) < 3:
        print("Error: User ID must be at least 3 characters long")
        sys.exit(1)
    if len(user_id_str) > 64:
        print("Error: User ID must be at most 64 characters long")
        sys.exit(1)
    
    # Get server list
    server_infos = []
    if args.servers:
        print("📋 Using manually specified servers...")
        server_urls = [url.strip() for url in args.servers.split(",")]
        print(f"   Servers specified: {len(server_urls)}")
        for i, url in enumerate(server_urls, 1):
            print(f"   {i}. {url}")
        
        # Get public keys directly from each server via GetServerInfo
        print("   🔍 Querying servers for public keys...")
        server_infos = []
        for url in server_urls:
            try:
                # Create a basic client to call GetServerInfo
                basic_client = OpenADPClient(url)
                server_info = basic_client.get_server_info()
                
                # Extract public key from server info
                public_key = ""
                if isinstance(server_info, dict) and "noise_nk_public_key" in server_info:
                    noise_key = server_info["noise_nk_public_key"]
                    if noise_key:
                        public_key = "ed25519:" + noise_key
                
                server_infos.append(ServerInfo(
                    url=url,
                    public_key=public_key,
                    country="Unknown"
                ))
                
                key_status = "❌ No public key" if not public_key else "🔐 Public key available"
                print(f"   ✅ {url} - {key_status}")
            except Exception as e:
                print(f"   ⚠️  Failed to get server info from {url}: {e}")
                # Add server without public key as fallback
                server_infos.append(ServerInfo(
                    url=url,
                    public_key="",
                    country="Unknown"
                ))
    else:
        print(f"🌐 Discovering servers from registry: {args.servers_url}")
        
        # Try to get full server information including public keys
        try:
            server_infos = get_servers(args.servers_url)
            if not server_infos:
                raise Exception("No servers returned from registry")
            print(f"   ✅ Successfully fetched {len(server_infos)} servers from registry")
        except Exception as e:
            print(f"   ⚠️  Failed to fetch from registry: {e}")
            print("   🔄 Falling back to hardcoded servers...")
            server_infos = get_fallback_server_info()
            print(f"   Fallback servers: {len(server_infos)}")
        
        print("   📋 Server list with public keys:")
        for i, server in enumerate(server_infos, 1):
            key_status = "❌ No public key" if not server.public_key else "🔐 Public key available"
            print(f"      {i}. {server.url} [{server.country}] - {key_status}")
    
    if not server_infos:
        print("❌ Error: No servers available")
        sys.exit(1)
    
    # Encrypt the file
    try:
        encrypt_file(args.filename, password_str, user_id_str, server_infos, args.servers_url)
        print("✅ File encrypted successfully!")
    except Exception as e:
        print(f"❌ Encryption failed: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main() 

File: sdk/cpp/tools/openadp-encrypt.cpp
#include <openadp.hpp>
#include <openadp/debug.hpp>
#include <iostream>
#include <fstream>
#include <sstream>
#include <getopt.h>
#include <termios.h>
#include <unistd.h>

using namespace openadp;

void print_usage(const char* program_name) {
    std::cout << "Usage: " << program_name << " [OPTIONS]\n"
              << "\n"
              << "OpenADP File Encryption Tool\n"
              << "\n"
              << "Options:\n"
              << "  --file <path>          File to encrypt (required)\n"
              << "  --password <password>  Password for key derivation (will prompt if not provided)\n"
              << "  --user-id <id>         User ID for secret ownership (will prompt if not provided)\n"
              << "  --servers <urls>       Comma-separated list of server URLs (optional)\n"
              << "  --servers-url <url>    URL to scrape for server list (default: https://servers.openadp.org/api/servers.json)\n"
              << "  --debug                Enable debug mode for deterministic operations\n"
              << "  --version              Show version information\n"
              << "  --help                 Show this help message\n"
              << "\n"
              << "USER ID SECURITY:\n"
              << "    Your User ID uniquely identifies your secrets on the servers. It is critical that:\n"
              << "    • You use the same User ID for all your files\n"
              << "    • You keep your User ID private (anyone with it can overwrite your secrets)\n"
              << "    • You choose a unique User ID that others won't guess\n"
              << "    • You remember your User ID for future decryption\n"
              << "\n"
              << "    You can set the OPENADP_USER_ID environment variable to avoid typing it repeatedly.\n"
              << "\n"
              << "SERVER DISCOVERY:\n"
              << "    By default, the tool fetches the server list from servers.openadp.org/api/servers.json\n"
              << "    If the registry is unavailable, it falls back to hardcoded servers.\n"
              << "    Use --servers to specify your own server list and skip discovery.\n"
              << "\n"
              << "DEBUG MODE:\n"
              << "    When --debug is enabled, all cryptographic operations become deterministic\n"
              << "    for testing purposes. This should NEVER be used in production.\n"
              << "\n"
              << "EXAMPLES:\n"
              << "    # Encrypt a file using discovered servers (fetches from servers.openadp.org)\n"
              << "    " << program_name << " --file document.txt\n"
              << "\n"
              << "    # Encrypt using specific servers (skip discovery)\n"
              << "    " << program_name << " --file document.txt --servers \"https://server1.com,https://server2.com\"\n"
              << "\n"
              << "    # Use a different server registry\n"
              << "    " << program_name << " --file document.txt --servers-url \"https://my-registry.com\"\n"
              << "\n"
              << "    # Enable debug mode for testing\n"
              << "    " << program_name << " --file document.txt --debug\n"
              << "\n"
              << "    # Use environment variables to avoid prompts\n"
              << "    export OPENADP_PASSWORD=\"mypassword\"\n"
              << "    export OPENADP_USER_ID=\"myuserid\"\n"
              << "    " << program_name << " --file document.txt\n"
              << "\n"
              << "The encrypted file will be saved as <filename>.enc\n";
}

std::string read_password() {
    std::cout << "Enter password: ";
    std::cout.flush();
    
    // Disable echo
    struct termios old_termios, new_termios;
    tcgetattr(STDIN_FILENO, &old_termios);
    new_termios = old_termios;
    new_termios.c_lflag &= ~ECHO;
    tcsetattr(STDIN_FILENO, TCSANOW, &new_termios);
    
    std::string password;
    std::getline(std::cin, password);
    
    // Restore echo
    tcsetattr(STDIN_FILENO, TCSANOW, &old_termios);
    
    std::cout << std::endl;
    return password;
}

std::string read_user_id() {
    std::cout << "Enter User ID: ";
    std::cout.flush();
    
    std::string user_id;
    std::getline(std::cin, user_id);
    
    return user_id;
}

int main(int argc, char* argv[]) {
    std::string input_file;
    std::string user_id;
    std::string password;
    std::string servers;
    std::string servers_url;
    bool debug_mode = false;
    
    static struct option long_options[] = {
        {"file", required_argument, 0, 'f'},
        {"password", required_argument, 0, 'p'},
        {"user-id", required_argument, 0, 'u'},
        {"servers", required_argument, 0, 's'},
        {"servers-url", required_argument, 0, 'r'},
        {"debug", no_argument, 0, 'd'},
        {"version", no_argument, 0, 'v'},
        {"help", no_argument, 0, 'h'},
        {0, 0, 0, 0}
    };
    
    int option_index = 0;
    int c;
    
    while ((c = getopt_long(argc, argv, "f:p:u:s:r:dvh", long_options, &option_index)) != -1) {
        switch (c) {
            case 'f':
                input_file = optarg;
                break;
            case 'p':
                password = optarg;
                break;
            case 'u':
                user_id = optarg;
                break;
            case 's':
                servers = optarg;
                break;
            case 'r':
                servers_url = optarg;
                break;
            case 'd':
                debug_mode = true;
                break;
            case 'v':
                std::cout << "OpenADP C++ Encrypt Tool v0.1.2\n";
                return 0;
            case 'h':
                print_usage(argv[0]);
                return 0;
            case '?':
                print_usage(argv[0]);
                return 1;
            default:
                break;
        }
    }
    
    // Set debug mode if requested
    if (debug_mode) {
        debug::set_debug(true);
    }
    
    // Validate required argument
    if (input_file.empty()) {
        std::cerr << "Error: Input file '--file' is required.\n\n";
        print_usage(argv[0]);
        return 1;
    }
    
    // Check if input file exists
    std::ifstream file_check(input_file);
    if (!file_check.good()) {
        std::cerr << "Error: Input file '" << input_file << "' not found.\n";
        return 1;
    }
    file_check.close();
    
    // Get user ID from environment or prompt
    if (user_id.empty()) {
        const char* env_user_id = std::getenv("OPENADP_USER_ID");
        if (env_user_id) {
            user_id = env_user_id;
        } else {
            user_id = read_user_id();
            if (user_id.empty()) {
                std::cerr << "Error: User ID cannot be empty.\n";
                return 1;
            }
        }
    } else {
        std::cerr << "⚠️  Warning: User ID provided via command line (visible in process list)\n";
    }
    
    // Get password from environment or prompt
    if (password.empty()) {
        const char* env_password = std::getenv("OPENADP_PASSWORD");
        if (env_password) {
            password = env_password;
        } else {
            password = read_password();
            if (password.empty()) {
                std::cerr << "Error: Password cannot be empty.\n";
                return 1;
            }
        }
    } else {
        std::cerr << "⚠️  Warning: Password provided via command line (visible in process list)\n";
    }
    
    // Generate output filename
    std::string output_file = input_file + ".enc";
    
    try {
        // Read input file
        Bytes plaintext = utils::read_file(input_file);
        
        // Create identity
        std::string device_id = "beast";
        std::string backup_id = "file://" + input_file;
        Identity identity(user_id, device_id, backup_id);
        
        // Handle --servers parameter by parsing comma-separated URLs
        EncryptResult result{Bytes(), Bytes()}; // Initialize with empty data
        if (!servers.empty()) {
            // Parse comma-separated server URLs into vector
            std::vector<ServerInfo> server_infos;
            std::stringstream ss(servers);
            std::string server_url;
            
            while (std::getline(ss, server_url, ',')) {
                // Trim whitespace
                server_url.erase(0, server_url.find_first_not_of(" \t"));
                server_url.erase(server_url.find_last_not_of(" \t") + 1);
                
                if (!server_url.empty()) {
                    server_infos.emplace_back(server_url);
                }
            }
            
            if (server_infos.empty()) {
                std::cerr << "Error: No valid servers found in --servers parameter.\n";
                return 1;
            }
            
            // Encrypt data using specific servers
            result = encrypt_data(plaintext, identity, password, 10, 0, server_infos);
        } else {
            // Encrypt data using server discovery
            result = encrypt_data(plaintext, identity, password, 10, 0, servers_url);
        }
        
        // Write encrypted data with embedded metadata (matching other SDKs)
        // Format: [metadata_length][metadata][nonce][encrypted_data]
        
        // Parse the existing metadata to extract components and rebuild in proper format
        std::string result_metadata_str = utils::bytes_to_string(result.metadata);
        nlohmann::json existing_metadata = nlohmann::json::parse(result_metadata_str);
        
        // Extract AES-GCM components from existing metadata
        Bytes nonce = utils::base64_decode(existing_metadata["nonce"].get<std::string>());
        Bytes tag = utils::base64_decode(existing_metadata["tag"].get<std::string>());
        
        // Create new metadata JSON matching other SDKs format (without embedded crypto data)
        // This must match exactly what was used as AAD during encryption
        nlohmann::json metadata;
        metadata["auth_code"] = existing_metadata["auth_code"];
        metadata["backup_id"] = existing_metadata["backup_id"];
        metadata["device_id"] = existing_metadata["device_id"];
        metadata["servers"] = existing_metadata["servers"];
        metadata["threshold"] = existing_metadata["threshold"];
        metadata["user_id"] = existing_metadata["user_id"];
        metadata["version"] = "1.0";
        
        std::string metadata_str = metadata.dump();
        std::vector<uint8_t> metadata_bytes(metadata_str.begin(), metadata_str.end());
        
        // Prepare final output: [metadata_length][metadata][nonce][encrypted_data]
        std::vector<uint8_t> output_data;
        
        // Write metadata length (4 bytes, little endian)
        uint32_t metadata_len = static_cast<uint32_t>(metadata_bytes.size());
        output_data.push_back(metadata_len & 0xFF);
        output_data.push_back((metadata_len >> 8) & 0xFF);
        output_data.push_back((metadata_len >> 16) & 0xFF);
        output_data.push_back((metadata_len >> 24) & 0xFF);
        
        // Write metadata
        output_data.insert(output_data.end(), metadata_bytes.begin(), metadata_bytes.end());
        
        // Write nonce (12 bytes)
        output_data.insert(output_data.end(), nonce.begin(), nonce.end());
        
        // Write encrypted data (ciphertext + auth tag combined)
        output_data.insert(output_data.end(), result.ciphertext.begin(), result.ciphertext.end());
        output_data.insert(output_data.end(), tag.begin(), tag.end());
        
        utils::write_file(output_file, output_data);
        
        // Get file sizes for display
        std::ifstream input_stream(input_file, std::ios::binary | std::ios::ate);
        std::streamsize input_size = input_stream.tellg();
        input_stream.close();
        
        std::ifstream output_stream(output_file, std::ios::binary | std::ios::ate);
        std::streamsize output_size = output_stream.tellg();
        output_stream.close();
        
        // Display results in the same format as other SDKs
        std::cout << "📁 Input:  " << input_file << " (" << input_size << " bytes)\n";
        std::cout << "📁 Output: " << output_file << " (" << output_size << " bytes)\n";
        std::cout << "🔐 Encryption: AES-GCM\n";
        std::cout << "✅ File encrypted successfully!\n";
        
        return 0;
        
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
} 